lib/axon/shared.ex

```diff
@@ -87,59 +85,66 @@ defmodule Axon.Shared do
   defn fulls_like(params, value) do
     transform(
       params,
-      &deep_new(&1, fn {k, x} ->
-        {k, Axon.Initializers.full(value, shape: Nx.shape(x))}
-      end)
+      &deep_new(&1, fn x -> Axon.Initializers.full(value, shape: Nx.shape(x)) end)
     )
   end
 
   @doc """
   Deep merges two possibly nested maps, applying fun to leaf values.
   """
   def deep_merge(left, right, fun) do
-    Map.merge(left, right, &deep_resolve(&1, &2, &3, fun))
+    {merged, []} = Nx.Container.traverse(left, leaves(right), &recur_merge(&1, &2, fun))
```

Comment:
Perhaps we should have a case and when the second tuple element is not an empty list, we raise saying the shapes are not compatible.